### ISSUE TITLE
MPS lgamma function changes results when using broadcasting

### ISSUE THREAD

**Issue by canergen** [pytorch/pytorch#132605]
### 🐛 Describe the bug

When broadcasting a tensor to another shape, torch.lgamma on MPS yields different results for the same input value (when copying back and forth between CPU and MPS the correct values are yield):

```
from torch.distributions.utils import broadcast_all
import torch

b = torch.full((3, 10), 1., device='mps')
c = torch.full((3, 1), 1., device='mps')
b, c = broadcast_all(b, c)
torch.lgamma(b), torch.lgamma(c), torch.lgamma(b.to('cpu').to('mps'))
```
yields:
```
(tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], device='mps:0'),
 tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., inf, inf, inf, inf, inf],
         [inf, inf, inf, inf, inf, inf, inf, inf, inf, inf]], device='mps:0'),
 tensor([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], device='mps:0'))
```

### Versions

Collecting environment information...
PyTorch version: 2.5.0.dev20240803
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:34:54) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-13.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] numpyro==0.15.2
[pip3] pytorch-lightning==2.3.3
[pip3] torch==2.5.0.dev20240803
[pip3] torchaudio==2.4.0.dev20240803
[pip3] torchmetrics==1.4.1
[pip3] torchvision==0.20.0.dev20240803
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] numpyro                   0.15.2                   pypi_0    pypi
[conda] pytorch                   2.5.0.dev20240803        py3.11_0    pytorch-nightly
[conda] pytorch-lightning         2.3.3                    pypi_0    pypi
[conda] torchaudio                2.4.0.dev20240803       py311_cpu    pytorch-nightly
[conda] torchmetrics              1.4.1                    pypi_0    pypi
[conda] torchvision               0.20.0.dev20240803       py311_cpu    pytorch-nightly

**Comment by skotapati** [comment]
Hi @canergen thanks for reporting this issue. I've been able to repro it locally, stay tuned for further updates

**Comment by canergen** [comment]
Great. Any idea already whether this is a more general problem with broadcasting (or view) or specific to lgamma? It would be easy in our codebase to fix the lgamma by using repeat.

**Comment by canergen** [comment]
@skotapati any news? Thanks

**Comment by skotapati** [comment]
Hi @canergen, sorry I missed your earlier question. This issue is more of a problem with broadcasting in mps than an issue with lgamma specifically. I'm still working on a fix, but if you can workaround the issue using repeat then that should work as a temporary fix

**Comment by canergen** [comment]
Thank you. To conclude, it's a more general problem with broadcasting in MPS. So simply removing broadcasting for lgamma will likely create issues with other uses of broadcasting (that don't lead to None errors).

**Comment by canergen** [comment]
@skotapati We made progress updating our code base to use MPS. However, this wrong behavior is not unique to lgamma and e.g. BatchNorm layers also use wrong array access in [BatchNorm(x_slice) for x_slice in x]. All of these errors are silent but just output the wrong result. This is significantly reducing my confidence in using torch MPS. It would be very helpful to at least print warnings if tensor access is insecure. Depending on the location in the close x_slice.clone() or a more lazy x_slice.contiguous() are sufficient to output the correct result. It would be straightforward to do these changes, if we would be able to track all of these issues. Currently, it's a lot of guessing and reverse engineering.

**Comment by skotapati** [comment]
Hi @canergen thanks for the update. I'm currently working on some testing updates for mps to catch similar silent correctness issues: https://github.com/pytorch/pytorch/pull/142202. If such tests don't exist already, I'll make sure to add cases which cover broadcasting, and add warnings for the relevant ops
